File: Stacks & Queues/10. Max Area in Histogram/fatimahadnan230--Q10. Max Area in Histogram.py
class Solution:
    def largestRectangleArea(self, heights: list[int]) -> int:
        stack = [] 
        max_area = 0
        # we add a 0 to the end to make sure the stack clears at the finish
        heights.append(0) 
        
        for i, h in enumerate(heights):
            # while current height is less than the height at the stack top
            while stack and heights[stack[-1]] >= h:
                height = heights[stack.pop()]
                # if stack is empty, it means this was the shortest bar so far
                width = i if not stack else i - stack[-1] - 1
                max_area = max(max_area, height * width)
            stack.append(i)
            
        # putting the list back to its original state
        heights.pop() 
        return max_area

File: Stacks & Queues/10. Max Area in Histogram/fatimahadnan230--Q10. Max Area in Histogram.cpp
#include <vector>
#include <stack>
#include <algorithm>

using namespace std;

class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<int> st;
        int maxArea = 0;
        int n = heights.size();
        
        for (int i = 0; i <= n; i++) {
            // treat the "past the end" element as height 0
            int currentHeight = (i == n) ? 0 : heights[i];
            
            while (!st.empty() && heights[st.top()] >= currentHeight) {
                int h = heights[st.top()];
                st.pop();
                
                int w = st.empty() ? i : i - st.top() - 1;
                maxArea = max(maxArea, h * w);
            }
            st.push(i);
        }
        
        return maxArea;
    }
};